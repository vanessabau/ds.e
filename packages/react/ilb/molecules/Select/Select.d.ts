import React from "react";
interface SelectOption {
    label: string;
    value: string;
}
interface SelectProps {
    onOptionSelected?(option: any, optionIndex: number): void;
    options?: SelectOption[];
    label?: string;
}
declare const Select: React.FC<SelectProps>;
export default Select;
