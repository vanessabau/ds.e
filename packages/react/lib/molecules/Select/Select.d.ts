import React from "react";
interface SelectOption {
    label: string;
    value: string;
}
interface RenderOptionProps {
    isSelected: boolean;
    option: SelectOption;
    getOptionRecommendedProps: (overrideProps?: Object) => Object;
}
interface SelectProps {
    onOptionSelected?(option: any, optionIndex: number): void;
    options?: SelectOption[];
    label?: string;
    renderOption?: (props: RenderOptionProps) => React.ReactNode;
}
declare const Select: React.FC<SelectProps>;
export default Select;
