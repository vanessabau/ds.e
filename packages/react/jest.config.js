module.exports = {
  testEnvironment: "jsdom",
  roots: ["<rootDir>/src"],
  //testRegex: "(/.*\\.test\\.(ts|tsx)$",
  setupFilesAfterEnv: ["<rootDir>/jest.setup.ts"],
};
