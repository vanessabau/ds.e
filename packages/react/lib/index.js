export { default as Button } from './atoms/Button/Button.js';
export { default as NewButton } from './atoms/NewButton/NewButton.js';
export { default as Color } from './atoms/Color/Color.js';
export { default as Margin } from './atoms/Margin/Margin.js';
export { default as Text } from './atoms/Text/Text.js';
export { default as Select } from './molecules/Select/Select.js';
export { default as Spacing } from './foundation/Spacing.js';
//# sourceMappingURL=index.js.map
