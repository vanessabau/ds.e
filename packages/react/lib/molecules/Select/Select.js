import React, { useState, useRef, useEffect, createRef } from 'react';
import Text from '../../atoms/Text/Text.js';

const getNextOptionIndex = (currentIndex, options) => {
    console.log(currentIndex);
    if (currentIndex === null) {
        return 0;
    }
    if (currentIndex === options.length - 1) {
        return 0;
    }
    return currentIndex + 1;
};
const getPreviousOptionIndex = (currentIndex, options) => {
    if (currentIndex === null) {
        return options.length - 1;
    }
    if (currentIndex === 0) {
        return 0;
    }
    return currentIndex - 1;
};
const Select = ({ options = [], label = "Please select an option", onOptionSelected,
//renderOption,
 }) => {
    const [isOpen, setIsOpen] = useState(false);
    const [selectedIndex, setSelectedIndex] = useState(null);
    const [highlightedIndex, setHighlightedIndex] = useState(null);
    const labelRef = useRef(null);
    const [optionRefs, setOptionRefs] = useState([]);
    const [overlayTop, setOverlayTop] = useState(0);
    const handleOptionSelected = (option, optionIndex) => {
        setIsOpen((prev) => !prev);
        onOptionSelected?.(option, optionIndex);
        setSelectedIndex(optionIndex);
        console.log(selectedIndex);
    };
    const onLabelClick = () => {
        setIsOpen((prev) => !prev);
    };
    useEffect(() => {
        setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
    }, [labelRef.current?.offsetHeight]);
    let selectedOption = null;
    if (selectedIndex !== null) {
        selectedOption = options[selectedIndex];
    }
    const highlightOption = (optionIndex) => {
        setHighlightedIndex(optionIndex);
    };
    const onButtonKeyDown = (event) => {
        event.preventDefault();
        console.log(event);
        // space is not working
        if (event.key === "Enter" ||
            event.key === "Space" ||
            event.key === "ArrowDown")
            setIsOpen(true);
        // Set focus on list item
        highlightOption(0);
    };
    useEffect(() => {
        setOptionRefs(options.map((_) => createRef()));
    }, [options.length]);
    useEffect(() => {
        if (highlightedIndex !== null && isOpen) {
            const ref = optionRefs[highlightedIndex];
            if (ref && ref.current) {
                ref.current.focus();
            }
        }
    }, [isOpen, highlightedIndex]);
    const onOptionKeyDown = (event) => {
        console.log(event);
        if (event.key === "Escape") {
            setIsOpen(false);
            return;
        }
        if (event.key === "ArrowDown") {
            highlightOption(getNextOptionIndex(highlightedIndex, options));
        }
        if (event.key === "ArrowUp") {
            highlightOption(getPreviousOptionIndex(highlightedIndex, options));
        }
        if (event.key === "Enter") {
            handleOptionSelected(options[highlightedIndex], highlightedIndex);
        }
    };
    return (React.createElement("div", { className: "dse-select" },
        React.createElement("button", { ref: labelRef, className: "dse-select__label", onClick: () => onLabelClick(), "aria-haspopup": true, "aria-expanded": isOpen ? true : undefined, "aria-controls": "dse-select-list", onKeyDown: onButtonKeyDown },
            React.createElement(Text, null, selectedOption === null ? label : selectedOption.label),
            React.createElement("svg", { className: `dse-select__caret ${isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"}`, width: "1rem", height: "1rem", xmlns: "http://www.w3.org/2000/svg", fill: "none", viewBox: "0 0 24 24", strokeWidth: 1.5, stroke: "currentColor" },
                React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M19.5 8.25l-7.5 7.5-7.5-7.5" }))),
        isOpen && (React.createElement("ul", { id: "dse-select-list", role: "menu", style: { top: overlayTop }, className: "dse-select__overlay" }, options.map((option, index) => {
            // selectedIndex is the same as optionIndex
            const isSelected = index === selectedIndex;
            const isHighlighted = highlightedIndex === selectedIndex;
            const ref = optionRefs[index];
            // const renderOptionProps = {
            //   option,
            //   isSelected,
            //   getOptionRecommendedProps: (overrideProps = {}) => {
            //     return {
            //       ref,
            //       onKeyDown: onOptionKeyDown,
            //       tabIndex: isHighlighted ? -1 : 0,
            //       onMouseEnter: () => highlightOption(selectedIndex),
            //       onMouseLeave: () => highlightOption(null),
            //       className: `dse-select__option
            //      ${isSelected ? "dse-select__option--selected" : ""}
            //      ${isHighlighted ? "dse-select__option--highlighted" : ""}`,
            //       key: option.value,
            //       onClick: handleOptionSelected(option, index),
            //       ...overrideProps,
            //     };
            //   },
            // };
            // if (renderOption) {
            //   return renderOption(renderOptionProps);
            // }
            return (React.createElement("li", { ref: ref, role: "menuitemradio", "aria-label": option.label, "aria-checked": isSelected ? true : undefined, className: `dse-select__option ${isSelected ? "dse-select__option--selected" : ""}`, key: option.value, onClick: () => {
                    handleOptionSelected(option, index);
                }, onKeyDown: onOptionKeyDown, tabIndex: isHighlighted ? -1 : 0, onMouseEnter: () => highlightOption(selectedIndex), onMouseLeave: () => highlightOption(null) },
                React.createElement(Text, null, option.label),
                isSelected ? (React.createElement("svg", { width: "1rem", height: "1rem", fill: "none", stroke: "currentColor", strokeWidth: 1.5, viewBox: "0 0 24 24", xmlns: "http://www.w3.org/2000/svg", "aria-hidden": "true" },
                    React.createElement("path", { strokeLinecap: "round", strokeLinejoin: "round", d: "M4.5 12.75l6 6 9-13.5" }))) : null));
        })))));
};

export { Select as default };
//# sourceMappingURL=Select.js.map
