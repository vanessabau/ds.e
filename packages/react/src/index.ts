import Button from "./atoms/Button";
import NewButton from "./atoms/NewButton";
import Color from "./atoms/Color/Color";
import Margin from "./atoms/Margin";
import Text from "./atoms/Text";
import Select from "./molecules/Select";

export { Button, NewButton, Color, Margin, Text, Select };
