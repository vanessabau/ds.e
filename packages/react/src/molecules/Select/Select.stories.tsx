import React from "react";
import Select from "./Select";

// css
import "@ds.e/scss/lib/Select.css";

const options = [
  {
    label: "Strict Black",
    value: "strict-black",
  },
  {
    label: "Heavenly Green",
    value: "heavenly-green",
  },
  {
    label: "Sweet Pink",
    value: "pink",
  },
];

// Group components by adding to the title
export default {
  title: "Molecules | Select",
  component: Select,
};

export const Common = () => <Select options={options} />;

export const CustomLabel = () => (
  <Select options={options} label="Select a Color" />
);

// Provide Developer-helpful variants, look at the component interface to reflect variants

// export const RenderOption = () => (
//   <Select
//     options={options}
//     renderOption={({ getOptionRecommendedProps, option, isSelected }) => {
//       <span {...getOptionRecommendedProps}>
//         {option.label}
//         {isSelected ? "SELECTED!" : ""}
//       </span>;
//     }}
//   />
// );
