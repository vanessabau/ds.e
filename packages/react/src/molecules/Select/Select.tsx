import React, {
  createRef,
  KeyboardEventHandler,
  useEffect,
  useRef,
  useState,
} from "react";
import Text from "../../atoms/Text";

interface SelectOption {
  label: string;
  value: string;
}

interface RenderOptionProps {
  isSelected: boolean;
  option: SelectOption;
  getOptionRecommendedProps: (overrideProps?: Object) => Object;
}

interface SelectProps {
  onOptionSelected?(option: any, optionIndex: number): void;
  options?: SelectOption[];
  label?: string;
  renderOption?: (props: RenderOptionProps) => React.ReactNode;
}

const getNextOptionIndex = (
  currentIndex: number | null,
  options: Array<SelectOption>
) => {
  console.log(currentIndex);
  if (currentIndex === null) {
    return 0;
  }
  if (currentIndex === options.length - 1) {
    return 0;
  }
  return currentIndex + 1;
};

const getPreviousOptionIndex = (
  currentIndex: number | null,
  options: Array<SelectOption>
) => {
  if (currentIndex === null) {
    return options.length - 1;
  }
  if (currentIndex === 0) {
    return 0;
  }
  return currentIndex - 1;
};

const Select: React.FC<SelectProps> = ({
  options = [],
  label = "Please select an option",
  onOptionSelected,
  //renderOption,
}) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedIndex, setSelectedIndex] = useState<null | number>(null);
  const [highlightedIndex, setHighlightedIndex] = useState<null | number>(null);
  const labelRef = useRef<HTMLButtonElement>(null);
  const [optionRefs, setOptionRefs] = useState<
    React.RefObject<HTMLLIElement>[]
  >([]);
  const [overlayTop, setOverlayTop] = useState<number>(0);

  const handleOptionSelected = (option: SelectOption, optionIndex: number) => {
    setIsOpen((prev) => !prev);
    onOptionSelected?.(option, optionIndex);
    setSelectedIndex(optionIndex);
    console.log(selectedIndex);
  };

  const onLabelClick = () => {
    setIsOpen((prev) => !prev);
  };

  useEffect(() => {
    setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
  }, [labelRef.current?.offsetHeight]);

  let selectedOption = null;

  if (selectedIndex !== null) {
    selectedOption = options[selectedIndex];
  }

  const highlightOption = (optionIndex: number | null) => {
    setHighlightedIndex(optionIndex);
  };

  const onButtonKeyDown: KeyboardEventHandler = (event) => {
    event.preventDefault();
    console.log(event);
    // space is not working
    if (
      event.key === "Enter" ||
      event.key === "Space" ||
      event.key === "ArrowDown"
    )
      setIsOpen(true);

    // Set focus on list item
    highlightOption(0);
  };

  useEffect(() => {
    setOptionRefs(options.map((_) => createRef<HTMLLIElement>()));
  }, [options.length]);

  useEffect(() => {
    if (highlightedIndex !== null && isOpen) {
      const ref = optionRefs[highlightedIndex];

      if (ref && ref.current) {
        ref.current.focus();
      }
    }
  }, [isOpen, highlightedIndex]);

  const onOptionKeyDown: KeyboardEventHandler = (event) => {
    console.log(event);
    if (event.key === "Escape") {
      setIsOpen(false);

      return;
    }

    if (event.key === "ArrowDown") {
      highlightOption(getNextOptionIndex(highlightedIndex, options));
    }

    if (event.key === "ArrowUp") {
      highlightOption(getPreviousOptionIndex(highlightedIndex, options));
    }

    if (event.key === "Enter") {
      handleOptionSelected(options[highlightedIndex!], highlightedIndex!);
    }
  };

  return (
    <div className="dse-select">
      <button
        ref={labelRef}
        className="dse-select__label"
        onClick={() => onLabelClick()}
        aria-haspopup={true}
        aria-expanded={isOpen ? true : undefined}
        aria-controls="dse-select-list"
        onKeyDown={onButtonKeyDown}
      >
        <Text>{selectedOption === null ? label : selectedOption.label}</Text>
        <svg
          className={`dse-select__caret ${
            isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"
          }`}
          width="1rem"
          height="1rem"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 24 24"
          strokeWidth={1.5}
          stroke="currentColor"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            d="M19.5 8.25l-7.5 7.5-7.5-7.5"
          />
        </svg>
      </button>

      {isOpen && (
        <ul
          id="dse-select-list"
          role="menu"
          style={{ top: overlayTop }}
          className="dse-select__overlay"
        >
          {options.map((option, index) => {
            // selectedIndex is the same as optionIndex
            const isSelected = index === selectedIndex;
            const isHighlighted = highlightedIndex === selectedIndex;

            const ref = optionRefs[index];

            // const renderOptionProps = {
            //   option,
            //   isSelected,
            //   getOptionRecommendedProps: (overrideProps = {}) => {
            //     return {
            //       ref,
            //       onKeyDown: onOptionKeyDown,
            //       tabIndex: isHighlighted ? -1 : 0,
            //       onMouseEnter: () => highlightOption(selectedIndex),
            //       onMouseLeave: () => highlightOption(null),
            //       className: `dse-select__option
            //      ${isSelected ? "dse-select__option--selected" : ""}
            //      ${isHighlighted ? "dse-select__option--highlighted" : ""}`,
            //       key: option.value,
            //       onClick: handleOptionSelected(option, index),
            //       ...overrideProps,
            //     };
            //   },
            // };

            // if (renderOption) {
            //   return renderOption(renderOptionProps);
            // }

            return (
              <li
                ref={ref}
                role="menuitemradio"
                aria-label={option.label}
                aria-checked={isSelected ? true : undefined}
                className={`dse-select__option ${
                  isSelected ? "dse-select__option--selected" : ""
                }`}
                key={option.value}
                onClick={() => {
                  handleOptionSelected(option, index);
                }}
                onKeyDown={onOptionKeyDown}
                tabIndex={isHighlighted ? -1 : 0}
                onMouseEnter={() => highlightOption(selectedIndex)}
                onMouseLeave={() => highlightOption(null)}
                // If you get the renderOptionProps working
                // {...renderOptionProps.getOptionRecommendedProps()}
              >
                <Text>{option.label}</Text>
                {isSelected ? (
                  <svg
                    width="1rem"
                    height="1rem"
                    fill="none"
                    stroke="currentColor"
                    strokeWidth={1.5}
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                    aria-hidden="true"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      d="M4.5 12.75l6 6 9-13.5"
                    />
                  </svg>
                ) : null}
              </li>
            );
          })}
        </ul>
      )}
    </div>
  );
};

export default Select;
