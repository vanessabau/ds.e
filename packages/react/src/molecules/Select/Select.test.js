import React from "react";
import Select from "./Select";

import { render, fireEvent } from "@testing-library/react";

const options = [
  {
    label: "Strict Black",
    value: "strict-black",
  },
  {
    label: "Heavenly Green",
    value: "heavenly-green",
  },
  {
    label: "Sweet Pink",
    value: "pink",
  },
];

test("renders all options passed to it", () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);

  fireEvent.click(getByTestId("DseSelectButton"));

  expect(getAllByRole("menuitemradio")).toHaveLength(options.length);
});
