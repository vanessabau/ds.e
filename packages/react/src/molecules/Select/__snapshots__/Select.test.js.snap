// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshop of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <div
        class="dse-text-md"
      >
        Strict Black
      </div>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.5 8.25l-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="dse-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <div
        class="dse-text-md"
      >
        Please select an option
      </div>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.5 8.25l-7.5 7.5-7.5-7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
